### PR TITLE
Orders

### PR DESCRIPTION
**Pull request by vovuk097**
Orders
State: Orders
Prototype: Orders

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -15,3 +15,5 @@
 
 # Ignore Byebug command history file.
 config/database.yml
+.DS_Store
```
Comment (vovuk097) [inline .gitignore]: 👍

File: app/services/order_states.rb [hunk 2]
```diff
@@ -0,0 +1,31 @@
+module OrderStates
```
Comment (ruslan-369) [inline app/services/order_states.rb:1]: FYI: https://github.com/aasm/aasm

File: app/assets/javascripts/orders/index.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
+//= require inspinia/sweetalert/sweetalert.min.js
+
+$(function() {
+  $('.swal').click(function () {
+    var thisSwal = $(this);
+    var index = $(this).data('index');
```
Comment (ruslan-369) [inline app/assets/javascripts/orders/index.js:6]: thisSwal.data('index');

File: app/assets/javascripts/orders/index.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
+//= require inspinia/sweetalert/sweetalert.min.js
```
Comment (ruslan-369) [inline app/assets/javascripts/orders/index.js:1]: what about using coffee?

File: app/concepts/order/_policy.rb [hunk 5]
```diff
@@ -0,0 +1,14 @@
+class Order
+  class Policy
+    attr_reader :model, :current_user
+
+    def initialize(current_user, model)
```
Comment (ruslan-369) [inline app/concepts/order/_policy.rb:5]: what about named args?

Comment (ruslan-369) [reply app/concepts/order/_policy.rb:5]: also better to rename current_user to user in this scope

File: app/concepts/order/cell/form.rb [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+      include ActionView::Helpers::FormOptionsHelper
+
+      property :contract
+
+      def show
+        render :form
+      end
+
+      def form_path
+        contract.model.new_record? ? orders_path : order_path
```
Comment (ruslan-369) [inline app/concepts/order/cell/form.rb:14]: add delegation for model to contract and use it everywhere in this class

File: app/concepts/order/cell/list.rb [hunk 7]
```diff
@@ -0,0 +1,24 @@
+class Order
+  class Cell
+    class List < Order::Cell
+      def show
+        render :list
+      end
+
+      def to_state
+        case model.state
```
Comment (ruslan-369) [inline app/concepts/order/cell/list.rb:9]: better to use mapping with hash instead of case statement

File: app/concepts/order/cell/list.rb [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+          'New Order'
+        when OrderStates::Canceled.to_s
+          'Canceled Order'
+        when OrderStates::Invoiced.to_s
+          'Invoiced Order'
+        end
+      end
+
+      def to_date(field)
+        field.strftime('%b %d, %Y - %I:%M %P') if field.present?
```
Comment (ruslan-369) [inline app/concepts/order/cell/list.rb:20]: mv format to const

Comment (ruslan-369) [reply app/concepts/order/cell/list.rb:20]: `field&.strftime('%b %d, %Y - %I:%M %P')` if field can't be blank string

File: app/concepts/order/contract/cancel.rb [hunk 9]
```diff
@@ -0,0 +1,11 @@
+class Order
+  module Contract
+    class Cancel < Reform::Form
+      validate :current_state
+
+      def current_state
+        errors.add(:order, 'You can`n cancel order') unless model.newmade?
```
Comment (ruslan-369) [inline app/concepts/order/contract/cancel.rb]: spelling error + I18n

File: app/concepts/order/contract/create.rb [hunk 10]
```diff
@@ -0,0 +1,13 @@
+class Order
+  module Contract
+    class Create < Reform::Form
+      properties :user_id, :name, :description, :item_ids
+      validates :name, presence: true
+      validate :current_state
+
+      def current_state
+        errors.add(:order, 'You can`n create new order') if model.user.orders.newmade.count >= 1
```
Comment (ruslan-369) [inline app/concepts/order/contract/create.rb]: `model.user.orders.newmade.count >= 1` => to another method
`1` => const
i18n

File: app/concepts/order/contract/reopen.rb [hunk 11]
```diff
@@ -0,0 +1,11 @@
+class Order
+  module Contract
+    class Reopen < Reform::Form
+      validate :current_state
+
+      def current_state
+        errors.add(:order, 'You can`n reopen order') if !model.canceled? || model.user.orders.newmade.take.present?
```
Comment (ruslan-369) [inline app/concepts/order/contract/reopen.rb]: `!model.canceled? || model.user.orders.newmade.take.present?` => another method

File: app/concepts/order/operation/cancel.rb [hunk 12]
```diff
@@ -0,0 +1,15 @@
+class Order
+  class Cancel < Trailblazer::Operation
```
Comment (ruslan-369) [inline app/concepts/order/operation/cancel.rb:2]: use ::BaseOperation

File: app/concepts/order/operation/cancel.rb [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+    include Dispatch
+    include Resolver
+    policy Order::Policy, :access?
+
+    include Model
+    model Order, :find
+    contract Contract::Cancel
+
+    def process(params)
+      validate({}) && model.cancel!
```
Comment (ruslan-369) [inline app/concepts/order/operation/cancel.rb:12]: why empty hash here?

File: app/concepts/order/operation/invoice.rb [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+    include Dispatch
+    include Resolver
+    policy Order::Policy, :access?
+
+    include Model
+    model Order, :find
+    contract Contract::Invoice
+
+    def process(params)
+      validate({}) && model.invoice!
```
Comment (ruslan-369) [inline app/concepts/order/operation/invoice.rb:12]: same here

File: app/concepts/order/operation/reopen.rb [hunk 15]
```diff
@@ -0,0 +1,15 @@
...
+    include Dispatch
+    include Resolver
+    policy Order::Policy, :access?
+
+    include Model
+    model Order, :find
+    contract Contract::Reopen
+
+    def process(params)
+      validate({}) && model.reopen!
```
Comment (ruslan-369) [inline app/concepts/order/operation/reopen.rb:12]: same

File: app/concepts/order/views/form.haml [hunk 16]
```diff
@@ -0,0 +1,34 @@
+.row.wrapper.border-bottom.white-bg.page-heading
```
Comment (ruslan-369) [inline app/concepts/order/views/form.haml:1]: use partials

File: app/concepts/order/views/list.haml [hunk 17]
```diff
@@ -0,0 +1,24 @@
+%tr
+  %td= link_to name, edit_order_path(model)
+  %td= to_state
+  %td= to_date created_at
+  %td.width-50
+    = form_tag order_path(model), method: :delete, 'data-index': model.id, class: 'swal-form-js display-none' do
+      .clearfix
+    %i.fa.fa-trash.swal.btn.btn-danger{ 'data-index': model.id, 'data-placement': 'left', 'data-toggle': 'delete-subscription-tooltip', title: I18n.t('action.destroy') }
+      = I18n.t('action.destroy')
```
Comment (ruslan-369) [inline app/concepts/order/views/list.haml:9]: use `t('action.destroy')` without prepended `I18n`

File: app/controllers/application_controller.rb [hunk 18]
```diff
@@ -3,6 +3,10 @@ class ApplicationController < ActionController::Base
   # For APIs, you may want to use :null_session instead.
   protect_from_forgery with: :exception
 
+  rescue_from ActiveRecord::RecordNotFound do
+    redirect_to root_url, alert: t('error_messages.record_not_found')
```
Comment (ruslan-369) [inline app/controllers/application_controller.rb]: root_path or try `:root` symbol

File: app/controllers/application_controller.rb [hunk 19]
```diff
@@ -23,4 +27,8 @@ def render_options
   def process_params!(params)
     params[:current_user] = current_user
   end
+
+  def should_authenticate
+    redirect_to root_url, alert: t('devise_views.signin.header') unless current_user.present?
```
Comment (ruslan-369) [inline app/controllers/application_controller.rb]: 1) same as above
2) use devise method `#user_signed_in?`

File: app/controllers/orders_controller.rb [hunk 20]
```diff
@@ -0,0 +1,77 @@
...
+
+  def invoice
+    run Order::Invoice do
+      return redirect_to orders_path, notice: t('successfull_messages.updated', model: t('activerecord.models.order'))
+    end
+    redirect_to orders_path, alert: @operation.errors.full_messages
+  end
+
+  def clone
+    OrderPrototype::DupPrototype.new(Order.find(params[:id])).clone
```
Comment (ruslan-369) [inline app/controllers/orders_controller.rb]: move `Order.find(params[:id])` to another method

File: app/controllers/orders_controller.rb [hunk 21]
```diff
@@ -0,0 +1,77 @@
...
+    end
+    flash[:alert] = @operation.errors.full_messages
+    render_form
+  end
+
+  def update
+    run Order::Update do
+      flash[:notice] = t('successfull_messages.updated', model: t('activerecord.models.order'))
+    end
+    flash[:alert] = @operation.errors.full_messages if @operation.errors.any?
```
Comment (ruslan-369) [inline app/controllers/orders_controller.rb:31]: is there need to check errors presence?

File: app/helpers/application_helper.rb [hunk 22]
```diff
@@ -7,4 +7,18 @@ def active_tab(controller, action = nil)
   def user_policy
     @user_policy ||= UserPolicy.new(current_user)
   end
+
+  def trash_link
```
Comment (ruslan-369) [inline app/helpers/application_helper.rb]: mv to orders_helper.rb

File: app/models/order.rb [hunk 23]
```diff
@@ -1,3 +1,27 @@
...
+  scope :canceled, -> { where(state: 'OrderStates::Canceled') }
+  scope :invoiced, -> { where(state: 'OrderStates::Invoiced') }
+
+  (OrderStates.constants - ['StateBase'.to_sym]).each do |type|
+    define_method "#{type.to_s.downcase}?" do
+      state_string_as_class == "OrderStates::#{type}".classify.constantize
+    end
+  end
+
+  alias orig_dup dup
```
Comment (ruslan-369) [inline app/models/order.rb:20]: what is reason for aliasing

File: app/models/order.rb [hunk 24]
```diff
@@ -1,3 +1,27 @@
 class Order < ApplicationRecord
+  include StatePattern::ActiveRecord
```
Comment (ruslan-369) [inline app/models/order.rb:2]: aasm

File: app/services/order_prototype.rb [hunk 25]
```diff
@@ -0,0 +1,37 @@
+module OrderPrototype
+  class PrototypeBase
+    def clone
+      raise NotImpelementedError.new
```
Comment (ruslan-369) [inline app/services/order_prototype.rb]: raise NotImpelementedError

File: app/models/order.rb [hunk 26]
```diff
@@ -1,3 +1,27 @@
...
+    define_method "#{type.to_s.downcase}?" do
+      state_string_as_class == "OrderStates::#{type}".classify.constantize
+    end
+  end
+
+  alias orig_dup dup
+
+  def dup_with_relations
+    copy = orig_dup
+    copy.items = items
```
Comment (ruslan-369) [inline app/models/order.rb:24]: plz double check if that will run query 
spontaneously and if it's behaviour u wanted to implement

File: app/views/layouts/partials/_header.html.haml [hunk 27]
```diff
@@ -5,3 +5,8 @@
         %i.fa.fa-bars
     %ul.nav.navbar-top-links.navbar-right
       %li
+        - if current_user.present?
```
Comment (ruslan-369) [inline app/views/layouts/partials/_header.html.haml]: user_signed_in?

**Review by ruslan-369 — DISMISSED** [review]

**Comment by nickKolya** [comment]
@adrenalin23 Good job, I agree with your comments, but we don't have time for its fixing, really

File: app/controllers/orders_controller.rb (hunk 21 above)
Comment (vovuk097) [inline app/controllers/orders_controller.rb]: yes

**Comment by vovuk097** [comment]
fixed comments

File: app/concepts/order/views/list.haml (hunk 17 above)
Comment (havrylyakar) [reply app/concepts/order/views/list.haml:9]: `t()` not avalible in cell views

File: app/assets/javascripts/orders/index.js (hunk 4 above)
Comment (havrylyakar) [reply app/assets/javascripts/orders/index.js:1]: we use simple js, not coffee

File: app/services/order_states.rb (hunk 2 above)
Comment (havrylyakar) [reply app/services/order_states.rb:1]: we dont have time do that

**Review by havrylyakar — APPROVED** [review]

File: app/services/order_states.rb (hunk 2 above)
Comment (ruslan-369) [reply app/services/order_states.rb:1]: it takes less time than implementation above and it's fyi)

File: app/assets/javascripts/orders/index.js (hunk 4 above)
Comment (ruslan-369) [reply app/assets/javascripts/orders/index.js:1]: why not? for ruby guy it's more convenient

**Pull request merged**